js/internal/modules/modules.go

```diff
@@ -35,7 +35,13 @@ var (
 func Get(name string) interface{} {
 	mx.RLock()
 	defer mx.RUnlock()
-	return modules[name]
+	mod := modules[name]
+	if i, ok := mod.(interface {
+		NewGlobalModule() interface{ NewModuleInstance() interface{} }
+	}); ok {
```

Comment:
This is a bit funky, but it's to avoid an import from `internal`, right? A comment would be nice to have.

Reply:
It was mostly because I didn't want to create a named interface in some package.
But if the approach is approved I am probably going to stuff it in `common` as I prefer the longer names (not just `New`) and this has potential for writing it wrong. 

I would argue some more "integration" tests will fix that, but it will be nicer to not need those for just checking that you provided the correct `interface`

Reply:
I can also move just the `HTTP` changes to a separate commit or PR they are kind of small really

Reply:
I'd definitely prefer named interfaces. Also, why the 2-level design here? What is the benefit of that compared to having just a single interface with `NewModuleInstance()`?

Reply:
Basically what I said in this [comment](https://github.com/loadimpact/k6/issues/1858#issuecomment-779891465)

Without the two levels, any "global" variables for each VU for a given module will need to be part of the one struct (or global variables). 

This makes less of sense for our production code as there we only have 1 instance of those, but for tests I would argue it will be nice to have multiple global module states without the need to clean them between tests ;). 

I am actually currently trying to move this to a separate PR with named interfaces and everything but the "global module instance per test" actually needs even more support in the modules registration framework as it technically needs to be made of instance per test and then that one instance to be used to create per VU instances of the module. That is currently not used by anything but the `data/SharedArray` can use that to not actually have to use `context` to store its data ;). 

I am pretty sure we can make it work without one of the levels(by just skipping the "root" struct) but I think this makes it more explicit and let us (if needed) have a shared state for all the global instances of a module - I can't currently figure out what that will be used for (neither in production code nor in the k6 tests) but :shrug: I prefer not to have to refactor this again just because we figure out a case.

Reply:
Yeah ... I also think I prefer to try to make that PR after this is merged ... either as is (in this parts of the code) or with small changes, as it turns out it will need to be slightly bigger and will probably also include slight changes to other modules in order to verify that what  we provide as interfaces is sufficient.

For example while trying to do it I figure out that we can technically give some vuState (unlike the lib.State) to each module instance - that can possibly have the `context` inside of it. That context needs to change between calls to the `iterations` so we can't just give it a `context`, so I am not particularly certain it is a great idea but might be what we use to get the context around instead of the current bridge :shrug:

Reply:
I am very confused, to be honest :sweat_smile: I don't see how you's be able to have both per-module and per-VU state with the current code, given that you currently _always_ make **both** a new "global" module and a new module instance for every VU that imports the module: https://github.com/loadimpact/k6/blob/71a349072b43f59ec716c1c72ee16d0a9c4c8fba/js/internal/modules/modules.go#L42

Besides, 
>  I can't currently figure out what that will be used for (neither in production code nor in the k6 tests) but :shrug:

Weren't you the one advocating for the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) before? :smile: If you want to merge this as it is, please remove one level of interfaces here and maybe add it again in the proper module. I also think that we might just add a separate interface in the future, if we ever need both per-module and per-VU storage in a module, but we can discuss that in the other PR you want to make.

Reply:
The point is that I want to make 
- per vu modules(NewModuleInstance) for this PR as that is what we need
- per test modules (NewGlobalModule) for things such as the `data/SharedArray`. This is currently not supported ... as well I haven't implemented it ;). Arguably it can be used for something in the HTTP module (or any other modules as well)
- the root module (naming stuff is hard :sob: ) is basically there so we can have something to generate the GlobalModule.

 Arguably it can just be a function that returns it not an interface (although as we know from [net/http](https://golang.org/pkg/net/http/#HandleFunc) those are isomorphic ;) ) which I guess will be ... easier to write by someone developing a module? But I also think that having 2 interfaces and three "types" RootModule, GlobalModule, ModuleInstance (have I mentioned naming stuff is hard) is easier to reason about :man_shrugging: 
- 
My current plan is that the modules registry will get the RootModule(whether a function or struct/interface) and then to get a module instance out you will need to instantiate the module *registry* which will call `NewGlobalModule` once and cache it for each module that is imported and call NewModuleInstance on that GlobalModule each time that is initialized. This (the module registry initialization) is IMO mostly useful for testing as otherwise if you decide to use the module registry in tests you will get 1 *shared* GlobalModule for all the test in a single *golang* package.

I guess I can drop some of it .. but if I am going to remake it, either way, remaking it in this PR seems pointless to me ;) We are not making this a public interface yet, or documentation it ...

Reply:
:+1: :man_shrugging: :sweat_smile: You are right, something like a global module instance would be better place for the `SharedArray` values than the current `lib.State` BS (we probably want to remove even more things from there!), and per-VU storage would be useful for things like NTLM and digest authentication caches.